src/authentication/authentication-hapi-plugin.ts

```diff
@@ -0,0 +1,154 @@
+import { AccessToken, authServerBaseUrlInjectSymbol, jwtOptionsInjectSymbol } from './types';
+
+interface UserId {
+  id: string;
+  idp: string;
+}
+
+@injectable()
+class AuthenticationHapiPlugin extends HapiPlugin {
+
+  public static injectSymbol = Symbol('authentication-hapi-plugin');
+  private readonly authenticationModule: AuthenticationModule;
+  private readonly authServerBaseUrl: string;
+  private readonly logger: Logger;
+  private readonly gitlab: GitlabClient;
+  private readonly options: auth.JWTStrategyOptions;
+  private teamTokenClaimKey: 'https://team_token' = 'https://team_token';
+  private subEmailClaimKey: 'https://sub_email' = 'https://sub_email';
+
+  constructor(
+    @inject(AuthenticationModule.injectSymbol) authenticationModule: AuthenticationModule,
+    @inject(authServerBaseUrlInjectSymbol) authServerBaseUrl: string,
+    @inject(loggerInjectSymbol) logger: Logger,
+    @inject(GitlabClient.injectSymbol) gitlab: GitlabClient,
+    @inject(jwtOptionsInjectSymbol) jwtOptions: auth.JWTStrategyOptions,
+    ) {
+    super({
+      name: 'authentication-plugin',
+      version: '1.0.0',
+    });
+    this.authenticationModule = authenticationModule;
+    this.authServerBaseUrl = authServerBaseUrl;
+    this.logger = logger;
+    this.gitlab = gitlab;
+    this.options = jwtOptions;
+  }
+
+  public async register(server: Hapi.Server, _options: Hapi.IServerOptions, next: () => void) {
+    await this.registerAuth(server);
+    server.route([{
+      method: 'GET',
+      path: '/signup/{teamToken}',
+      handler: {
+        async: this.signupHandler,
+      },
+      config: {
+        bind: this,
+        validate: {
+          params: {
+            teamToken: Joi.string().alphanum().length(8),
+          },
+        },
+      },
+    },
+    ]);
+    server.route([{
+      method: 'GET',
+      path: '/team',
+      handler: {
+        async: this.getTeamHandler,
+      },
+      config: {
+        bind: this,
+      },
+    }]);
+    next();
+  }
+
+  public async signupHandler(request: Hapi.Request, reply: Hapi.IReply) {
+    const credentials = request.auth.credentials as AccessToken;
+    const teamToken = credentials[this.teamTokenClaimKey];
+
+    reply(teamToken);
+  }
+
+  public async getTeamHandler(request: Hapi.Request, reply: Hapi.IReply) {
+    const credentials = request.auth.credentials as AccessToken;
+    const user = await getUserByEmail(credentials[this.subEmailClaimKey], this.gitlab);
+    const teams = getUserTeams(user.id, this.gitlab);
+    return reply(teams);
+  }
+
+  public async signToTeam(_uid: UserId, teamToken: string) {
+    return teamToken;
+  }
+
+  public async getTeam(uid: UserId) {
+    return uid.idp;
+  }
+
+  public async validateUser(
+    payload: AccessToken,
+    _request: any,
+    callback: (err: any, valid: boolean, credentials?: any) => void,
+  ) {
+    // NOTE: we have just a single team at this point so no further checks are necessary
```

Comment:
It's possible for any number of reasons for the signup redirect to have failed, in which case the above mentioned scenario would come to pass. 

If we had multiple teams these checks and more would need to be made anyway. Meanwhile any issues manifest only when the user wants to actually use the git repo. If the redirect failed however, the user doesn't have any credentials anyway so this would need to be fixed manually in any case.